### PR TITLE
Bulk Load CDK: Async part upload dispatch per part, not per write

### PR DESCRIPTION
**Pull request by johnny-schmidt**
### What
S3MultipartUpload dispatches work per-part instead of per-write.

The underlying issue comes from how we're presenting the S3 multipart upload process as an output stream. Workflow:
* start a multipart upload
* present an output stream to client in a closure
* client writes to the output stream
* data is accumulated in a buffer
* when we reach the configured part size, we kick off `uploadPart`
* when the closure completes, we complete the upload (data becomes visible in s3)

The problem is that the AWS Kotlin SDK uses coroutines, but `OutputStream` does not. So whenever a suspend function is triggered by a write, we have to bridge.

Naively, you'd just wrap everything in `runBlocking`, but if you do this with long-running tasks it can block other coroutines on the thread and lead to unpredictable behavior.

The original solution was to dispatch all work on channel and process it asynchronously. This is not ideal, because it imposed a cost per-write. However, it did not initially seem to impact performance (much?) because most writes were a byte array per row or formatted block.

CSV however calls `write` per-byte, so the cost was enough to trigger a 10x slowdown.

### How

The solution is
* remove all unnecessary suspend functions from the interface (the only thing that absolutely must be suspend is the API calls)
* do the buffer accumulation synchronously, and only enqueue the finished byte arrays
* process the byte arrays asynchronously as upload parts api calls

### Results
See this [sheet](https://docs.google.com/spreadsheets/d/11WkVCiPh0AiIJvARgrFTv4oyQRB7_qvpKYogdgfkpPE/edit?gid=0#gid=0) for details. TLDR:
* json: slight perf hit -> parity with the old CDK
* csv: 10x perf hit -> parity with the old CDK
* parquet: slight perf hit -> 2x *improvement* over the old CDK
* avro: unclear; perf on the new cdk relative to pre-bug 2x'ed; but the old CDK tests failed to finish in a timely fashion (probably due to transient conditions?) so I need to re-run them to get a good comparison

### CODE REVIEW

File: airbyte-integrations/connectors/destination-s3/build.gradle [hunk 1]
```diff
@@ -17,6 +17,21 @@ application {
 
     // uncomment and replace to run locally
     //applicationDefaultJvmArgs = ['-XX:+ExitOnOutOfMemoryError', '-XX:MaxRAMPercentage=75.0', '--add-opens', 'java.base/java.lang=ALL-UNNAMED']
+    applicationDefaultJvmArgs = [
+            '-XX:+ExitOnOutOfMemoryError', '-XX:MaxRAMPercentage=75.0'
+//            Uncomment to enable remote profiling:
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-s3/build.gradle:22]: curious, was this helpful?

Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-s3/build.gradle:22]: I figured it out locally before I ran the profiler, but the profiler helped confirm it, yes.

Specifically I could see the high # of allocations in the write method.

Comment (davinchia) [reply airbyte-integrations/connectors/destination-s3/build.gradle:22]: great. so you managed to figure out connecting to the K8s cluster + port forwarding and all?

File: airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3MultipartUpload.kt [hunk 2]
```diff
@@ -80,57 +86,56 @@ class S3MultipartUpload<T : OutputStream>(
...
+            if (isClosed.setOnce()) {
+                partQueue.close()
             }
         }
 
-        override fun flush() = runBlocking { workQueue.send { wrappingBuffer.flush() } }
+        override fun flush() = wrappingBuffer.flush()
 
-        override fun write(b: Int) = runBlocking {
-            workQueue.send {
```
Comment (davinchia) [inline airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3MultipartUpload.kt:92]: help me out while trying to understand where the previous bug was - is it here?

Comment (johnny-schmidt) [reply airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3MultipartUpload.kt:92]: Yes, specifically it was paying the cost per byte of some combination of
* de/allocating this lambda passed through send
* de/allocating the hidden lambda due to `send` being a suspend fun
* whatever additional overhead is imposed by making interacting with the channel thread-safe

Comment (johnny-schmidt) [reply airbyte-cdk/bulk/toolkits/load-s3/src/main/kotlin/io/airbyte/cdk/load/file/s3/S3MultipartUpload.kt:92]: the `write(b: ByteArray)` was also more costly than it needed to be, showing up as a slight perf hit on json/avro. But the big one was CSV.

**Review by tryangul — APPROVED** [review]

**Pull request merged**